File: scrapers/spotify_test.py
import requests
import time
import json
import os
import sys
import logging
from concurrent.futures import ThreadPoolExecutor, as_completed

# ---------- settings ----------
SCRAPERS_DIR = os.path.dirname(os.path.abspath(__file__))
PROJECT_ROOT = os.path.normpath(os.path.join(SCRAPERS_DIR, ".."))
CONFIG_FILE = os.path.join(PROJECT_ROOT, "config", "spotify_config.json")
OUT_DIR = os.path.join(PROJECT_ROOT, "data", "spotify")
MAX_WORKERS = 5      # charts fetched at the same time (polite)
RETRIES = 3          # tries per chart before giving up
PAUSE = 0.3          # gap between charts so we don't hammer Spotify

try:
    with open(CONFIG_FILE, encoding="utf-8") as f:
        CONFIG = json.load(f)
    LIMIT = CONFIG["limit"]
except FileNotFoundError:
    sys.exit(f"ERROR: config file not found: {CONFIG_FILE}")
except json.JSONDecodeError as e:
    sys.exit(f"ERROR: invalid JSON in {CONFIG_FILE}: {e}")
except KeyError as e:
    sys.exit(f"ERROR: missing key in {CONFIG_FILE}: {e}")

LOCK_FILE = os.path.join(PROJECT_ROOT, ".spotify.lock")
LOG_FILE  = os.path.join(PROJECT_ROOT, "logs", "spotify.log")
os.makedirs(os.path.join(PROJECT_ROOT, "logs"), exist_ok=True)
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(message)s",
    handlers=[logging.FileHandler(LOG_FILE), logging.StreamHandler()],
)
log = logging.getLogger(__name__)


def build_url(country, slug):
    return f"https://podcastcharts.byspotify.com/api/charts/{slug}?region={country}&limit={LIMIT}"


def clean_entry(entry, rank):
    uri = entry.get("showUri", "")
    spotify_id = uri.split(":")[-1] if uri else ""
    return {
        "rank": rank,
        "spotify_id": spotify_id,
        "name": entry["showName"],
        "publisher": entry["showPublisher"],
        "artwork": entry["showImageUrl"],
        "rank_move": entry["chartRankMove"],
        "url": f"https://open.spotify.com/show/{spotify_id}" if spotify_id else "",
    }


def fetch_and_save(country, chart):
    time.sleep(PAUSE)
    url = build_url(country, chart["slug"])
    for attempt in range(1, RETRIES + 1):
        try:
            r = requests.get(url, timeout=15)

            # This chart isn't offered for this country -> skip quietly, don't retry.
            if r.status_code == 404:
                return {"status": "skip", "country": country, "slug": chart["slug"]}
            r.raise_for_status()

            entries = r.json()
            if not entries:                                 # 200 but empty -> also skip
                return {"status": "skip", "country": country, "slug": chart["slug"]}

            shows = [clean_entry(e, i) for i, e in enumerate(entries, start=1)]

            chart_data = {
                "platform": "spotify",
                "country": country,
                "genre": chart["name"],
                "slug": chart["slug"],
                "fetched_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
                "count": len(shows),
                "shows": shows,
            }

            # One clean file per chart: data/spotify/{country}/{slug}.json
            folder = os.path.join(OUT_DIR, country)
            os.makedirs(folder, exist_ok=True)
            path = os.path.join(folder, f"{chart['slug']}.json")
            tmp = path + ".tmp"
            with open(tmp, "w", encoding="utf-8") as out:
                json.dump(chart_data, out, ensure_ascii=False, indent=2)
            os.replace(tmp, path)

            return {"status": "ok", "country": country, "slug": chart["slug"], "count": len(shows)}

        except Exception as err:
            if attempt == RETRIES:
                return {"status": "fail", "country": country, "slug": chart["slug"], "error": str(err)}
            time.sleep(2 * attempt)   # wait longer each retry


def main():
    try:
        import fcntl
        _lf = open(LOCK_FILE, "w")
        fcntl.flock(_lf, fcntl.LOCK_EX | fcntl.LOCK_NB)
    except ImportError:
        pass  # Windows dev environment — skip locking
    except OSError:
        sys.exit("ERROR: another instance is already running")

    jobs = [(c, g) for c in CONFIG["countries"] for g in CONFIG["genres"]]
    log.info(f"Trying {len(jobs)} charts into {OUT_DIR}/ ...")

    ok = skip = fail = 0
    with ThreadPoolExecutor(max_workers=MAX_WORKERS) as pool:
        futures = [pool.submit(fetch_and_save, c, g) for c, g in jobs]
        for fut in as_completed(futures):
            res = fut.result()
            if res["status"] == "ok":
                ok += 1
                log.info(f"ok   {res['country']}/{res['slug']}: {res['count']} shows")
            elif res["status"] == "skip":
                skip += 1
            else:
                fail += 1
                log.error(f"FAIL {res['country']}/{res['slug']}: {res['error']}")

    log.info(f"Done. {ok} ok, {skip} skipped (not offered), {fail} failed.")
    if fail:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scrapers/spotify_test.php
<?php

$scrapersDir = __DIR__;
$projectRoot = realpath($scrapersDir . '/..');
$configFile = $projectRoot . '/config/spotify_config.json';
$outDir = $projectRoot . '/data/spotify';

define('MAX_WORKERS', 5);
define('RETRIES', 3);
define('PAUSE_US', 300000); // 0.3 seconds in microseconds

// --- Load Config ---
if (!file_exists($configFile)) {
    die("ERROR: config file not found: {$configFile}\n");
}
$configJson = file_get_contents($configFile);
$config = json_decode($configJson, true);
if (json_last_error() !== JSON_ERROR_NONE) {
    die("ERROR: invalid JSON in {$configFile}: " . json_last_error_msg() . "\n");
}
if (!isset($config['limit']) || !isset($config['countries']) || !isset($config['genres'])) {
    die("ERROR: missing key in {$configFile}\n");
}
$limit = $config['limit'];

// --- Single Instance Locking ---
$lockFile = $projectRoot . '/.spotify.lock';
$lockHandle = fopen($lockFile, 'w');
if (!$lockHandle || !flock($lockHandle, LOCK_EX | LOCK_NB)) {
    die("ERROR: another instance is already running\n");
}

// --- Logging Setup ---
$logDir = $projectRoot . '/logs';
if (!is_dir($logDir)) mkdir($logDir, 0777, true);
$logFile = $logDir . '/spotify.log';

function logMsg($level, $message) {
    global $logFile;
    $time = date('Y-m-d H:i:s,000'); 
    $line = "{$time} {$level} {$message}\n";
    echo $line;
    file_put_contents($logFile, $line, FILE_APPEND);
}

// --- URL Builder ---
function build_url($country, $slug, $limit) {
    return "https://podcastcharts.byspotify.com/api/charts/{$slug}?region={$country}&limit={$limit}";
}

// --- Data Cleaner ---
function clean_entry($entry, $rank) {
    $uri = $entry['showUri'] ?? '';
    $parts = explode(':', $uri);
    $spotifyId = end($parts); // gets the last item after the colon
    
    return [
        "rank"       => $rank,
        "spotify_id" => $spotifyId ? $spotifyId : "",
        "name"       => $entry['showName'] ?? '',
        "publisher"  => $entry['showPublisher'] ?? '',
        "artwork"    => $entry['showImageUrl'] ?? '',
        "rank_move"  => $entry['chartRankMove'] ?? null,
        "url"        => $spotifyId ? "https://open.spotify.com/show/{$spotifyId}" : "",
    ];
}

// --- Initialize Jobs Queue ---
$jobs = [];
foreach ($config['countries'] as $c) {
    foreach ($config['genres'] as $g) {
        $jobs[] = [
            'country' => $c,
            'genre'   => $g,
            'attempt' => 1
        ];
    }
}

logMsg('INFO', "Trying " . count($jobs) . " charts into {$outDir}/ ...");

$okCount = 0;
$skipCount = 0;
$failCount = 0;

// --- Concurrency via curl_multi ---
$mh = curl_multi_init();
$activeHandles = [];
$jobIndex = 0;

$addJobToMulti = function($job) use (&$mh, &$activeHandles, $limit) {
    usleep(PAUSE_US);
    $ch = curl_init();
    $url = build_url($job['country'], $job['genre']['slug'], $limit);
    
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_TIMEOUT, 15);
    curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
    
    curl_multi_add_handle($mh, $ch);
    $activeHandles[(int)$ch] = ['ch' => $ch, 'job' => $job];
};

// 1. Prime the pump
while ($jobIndex < count($jobs) && count($activeHandles) < MAX_WORKERS) {
    $addJobToMulti($jobs[$jobIndex]);
    $jobIndex++;
}

// 2. The Execution Loop
do {
    $status = curl_multi_exec($mh, $active);
    if ($active) {
        curl_multi_select($mh);
    }

    // 3. Process completed requests
    while ($info = curl_multi_info_read($mh)) {
        $ch = $info['handle'];
        $handleData = $activeHandles[(int)$ch];
        $job = $handleData['job'];
        
        $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
        $content = curl_multi_getcontent($ch);
        $curlError = curl_error($ch);

        // Cleanup handle
        curl_multi_remove_handle($mh, $ch);
        curl_close($ch);
        unset($activeHandles[(int)$ch]);

        $jobStatus = "fail"; // default to fail, override if ok or skip
        $errorMsg = "";
        $shows = []; // initialized to prevent Intelephense warnings

        // Check for 404 (Chart not offered in this country)
        if ($httpCode == 404) {
            $jobStatus = "skip";
        } 
        // Process successful 2xx responses
        elseif ($httpCode >= 200 && $httpCode < 300) {
            $entries = json_decode($content, true);
            
            // If the JSON is completely empty (e.g., []), skip it
            if (empty($entries)) {
                $jobStatus = "skip";
            } else {
                $rank = 1;
                foreach ($entries as $e) {
                    $shows[] = clean_entry($e, $rank++);
                }

                $chartData = [
                    "platform"   => "spotify",
                    "country"    => $job['country'],
                    "genre"      => $job['genre']['name'],
                    "slug"       => $job['genre']['slug'],
                    "fetched_at" => gmdate("Y-m-d\TH:i:s\Z"),
                    "count"      => count($shows),
                    "shows"      => $shows
                ];

                $folder = $outDir . '/' . $job['country'];
                if (!is_dir($folder)) mkdir($folder, 0777, true);
                
                $path = $folder . '/' . $job['genre']['slug'] . '.json';
                $tmp = $path . '.tmp';
                
                file_put_contents($tmp, json_encode($chartData, JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT));
                rename($tmp, $path); // Atomic replace

                $jobStatus = "ok";
            }
        } else {
            $errorMsg = $curlError ? $curlError : "HTTP Code $httpCode";
        }

        // 4. Handle Results and Retries
        if ($jobStatus === "ok") {
            $okCount++;
            logMsg('INFO', "ok   {$job['country']}/{$job['genre']['slug']}: " . count($shows) . " shows");
        } elseif ($jobStatus === "skip") {
            $skipCount++;
            // We do not log skips to keep the console clean, matching the Python script's behavior
        } else {
            // It failed. Check if we have retries left.
            if ($job['attempt'] < RETRIES) {
                $job['attempt']++;
                // Push back into the array to be retried naturally later
                $jobs[] = $job; 
            } else {
                $failCount++;
                logMsg('ERROR', "FAIL {$job['country']}/{$job['genre']['slug']}: {$errorMsg}");
            }
        }

        // 5. Keep the queue full
        if ($jobIndex < count($jobs)) {
            $addJobToMulti($jobs[$jobIndex]);
            $jobIndex++;
        }
    }
} while ($active || count($activeHandles) > 0);

curl_multi_close($mh);

logMsg('INFO', "Done. {$okCount} ok, {$skipCount} skipped (not offered), {$failCount} failed.");

// Release the instance lock
flock($lockHandle, LOCK_UN);
fclose($lockHandle);

// Exit code for run_pipeline tracking
if ($failCount > 0) {
    exit(1);
}